test/src/vfs_test_helper.h

```diff
@@ -0,0 +1,401 @@
+ * the S3 filesystem.
+ */
+class SupportedFsS3 : public SupportedFs {
+ public:
+  ~SupportedFsS3() = default;
+
+  /* ********************************* */
+  /*               API                 */
+  /* ********************************* */
+
+  /**
+   * Returns Status upon setting up the associated
+   * filesystem's configuration
+   *
+   * @param config Configuration parameters
+   * @return Status OK if successful
+   */
+  virtual Status prepare_config(tiledb_config_t* config, tiledb_error_t* error);
+
+  /**
+   * Creates bucket if does not exist
+   *
+   * @param ctx The TileDB context.
+   * @param vfs The VFS object.
+   * @return Status OK if successful
+   */
+  virtual Status init(tiledb_ctx_t* ctx, tiledb_vfs_t* vfs);
+
+  /**
+   * Removes bucket if exists
+   *
+   * @param ctx The TileDB context.
+   * @param vfs The VFS object.
+   * @return Status OK if successful
+   */
+  virtual Status close(tiledb_ctx_t* ctx, tiledb_vfs_t* vfs);
+
+  /**
+   * Get the name of the filesystem's directory
+   *
+   * @return string directory name
+   */
+  virtual std::string temp_dir();
+
+  /* ********************************* */
+  /*           ATTRIBUTES              */
+  /* ********************************* */
+
+  /** The directory prefix of the S3 filesystem. */
+  const std::string s3_prefix_ = "s3://";
+
+  /** The bucket name for the S3 filesystem. */
+  const std::string s3_bucket_ = s3_prefix_ + random_name("tiledb") + "/";
+
+  /** The directory name of the S3 filesystem. */
+  const std::string s3_temp_dir_ = s3_bucket_ + "tiledb_test/";
+
+  /** The directory name of the S3 filesystem. */
+  std::string temp_dir_ = s3_temp_dir_;
+};
+
+/**
+ * This class provides support for
+ * the HDFS filesystem.
+ */
+class SupportedFsHDFS : public SupportedFs {
+ public:
+  ~SupportedFsHDFS() = default;
+  /* ********************************* */
+  /*               API                 */
+  /* ********************************* */
+
+  /**
+   * No-op
+   *
+   * @param config Configuration parameters
+   * @return Status OK if successful
+   */
+  virtual Status prepare_config(tiledb_config_t* config, tiledb_error_t* error);
+
+  /**
+   * No-op
+   *
+   * @param ctx The TileDB context.
+   * @param vfs The VFS object.
+   * @return Status OK if successful
+   */
+  virtual Status init(tiledb_ctx_t* ctx, tiledb_vfs_t* vfs);
+
+  /**
+   * No-op
+   *
+   * @param ctx The TileDB context.
+   * @param vfs The VFS object.
+   * @return Status OK if successful
+   */
+  virtual Status close(tiledb_ctx_t* ctx, tiledb_vfs_t* vfs);
+
+  /**
+   * Get the name of the filesystem's directory
+   *
+   * @return string directory name
+   */
+  virtual std::string temp_dir();
+
+  /* ********************************* */
+  /*           ATTRIBUTES              */
+  /* ********************************* */
+
+  /** The directory name of the HDFS filesystem. */
+  const std::string hdfs_temp_dir_ = "hdfs:///tiledb_test/";
```

Comment:
Don't initialize member variables in the class definition. Initialize these in a constructor, preferably in the member-list initializer. Same elsewhere.